sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New language restrictions for device functions
+
+The following restriction, listed in section 5.4 of the core SYCL specification
+does not apply if this extension is supported by an implementation:
+
+> The odr-use of polymorphic classes and classes with virtual inheritance is
+> allowed. *However, no virtual member functions are allowed to be called in a
+> device function.*
+
+However, there are still some limitations of how virtual member functions can
+be used:
+
+- if an object is constructed in host code, calling a virtual member function
+  for that object in device code has undefined behavior;
+- if an object is constructed in device code on a device `A`, calling a virtual
+  member function for that object in host code, or on another device `B` has
+  undefined behavior;
+
+=== New properties
+
+Due to the indirect nature of virtual member functions, compiler may not be able
+to understand which exact virtual member function is being called (i.e. which
+class it belongs to).  Moreover, when code is distributed over several
+translation units, compiler may not be able to even see definitions of all
+virtual member functions which may be called in a kernel.
+
+Therefore, to provide a mechanism for an implementation to help detect virtual
+member functions, which are going to be used from kernels and enforce necessary
+restrictions on them, new compile-time-constant properties are proposed.
+
+[source,dpcpp]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  struct indirectly_callable_key {
+    template <typename SetId>
+    using value_t = property_value<indirectly_callable_key, SetId>;
+  };
+
+  struct calls_indirectly_key {
+    template <typename First = void, typename... SetIds>
+    using value_t = property_value<calls_indirectly_key, First, SetIds...>;
+  };
+
+  template <typename SetId = void>
+  inline constexpr indirectly_callable_key::value_t<SetId> indirectly_callable;
+
+  template <typename First = void, typename... Rest>
+  inline constexpr calls_indirectly_key::value_t<First, Rest...>
+      calls_indirectly;
+
+  template <>
+  struct is_property_key<indirectly_callable_key> : std::true_type {};
+  template <> struct is_property_key<calls_indirectly_key> : std::true_type {};
+}
+----
+
+Before describing those properties those properties in more detail, a couple of
+new terms are introduced to simplify the extension specification:
+
+Set of virtual member functions:: a group of virtual member functions which are
+defined with `indirectly_callable` property with the same value of the property
```

Comment:
```suggestion
defined with the `indirectly_callable` property and with the same value of the property
```

Reply:
Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783